lib/puppet/module/task.rb

```diff
@@ -33,54 +112,75 @@ def self.is_tasks_executable_filename?(name)
     end
 
     def self.tasks_in_module(pup_module)
-      Dir.glob(File.join(pup_module.tasks_directory, '*'))
+      task_files = Dir.glob(File.join(pup_module.tasks_directory, '*'))
         .keep_if { |f| is_tasks_filename?(f) }
-        .group_by { |f| task_name_from_path(f) }
-        .map { |task, files| new_with_files(pup_module, task, files) }
-    end
 
-    attr_reader :name, :module, :metadata_file, :files
+      module_executables = task_files.reject(&method(:is_tasks_metadata_filename?)).map.to_a
 
-    def initialize(pup_module, task_name, files, metadata_file = nil)
-      if !Puppet::Module::Task.is_task_name?(task_name)
-        raise InvalidName, _("Task names must start with a lowercase letter and be composed of only lowercase letters, numbers, and underscores")
+      tasks = task_files.group_by { |f| task_name_from_path(f) }
+
+      tasks.map do |task, executables|
+        new_with_files(pup_module, task, executables, module_executables)
       end
+    end
 
-      all_files = metadata_file.nil? ? files : files + [metadata_file]
-      all_files.each do |f|
-        if !f.start_with?(pup_module.tasks_directory)
-          msg = _("The file '%{path}' is not located in the %{module_name} module's tasks directory") %
-                       {path: f.to_s, module_name: pup_module.name}
+    attr_reader :name, :module, :metadata_file, :metadata, :files
 
-          # we can include some extra context for the log message:
-          Puppet.err(msg + " (#{pup_module.tasks_directory})")
-          raise InvalidFile, msg
-        end
+    # file paths must be relative to the modules task directory
+    def initialize(pup_module, task_name, files, metadata_file = nil, module_executables = nil)
+      if !Puppet::Module::Task.is_task_name?(task_name)
+        raise InvalidName, _("Task names must start with a lowercase letter and be composed of only lowercase letters, numbers, and underscores")
       end
 
       name = task_name == "init" ? pup_module.name : "#{pup_module.name}::#{task_name}"
 
       @module = pup_module
       @name = name
-      @metadata_file = metadata_file if metadata_file
+      @metadata_file = metadata_file
       @files = files
```

Comment:
What's the purpose of `@files` if we have `@module_executables`? It looks like the only thing we do with it is concat it with `@module_executables`, which I think should be a superset of `@files` already?

Reply:
yeah that is true now. I'll clean it up